### DISCUSSION TITLE
Need evolution and testing plan

### DISCUSSION THREAD

**Discussion by MikeSpreitzer** [kubestellar/kubestellar#444] [General]
We need a clearly documented plan for how we can make changes without causing breakage for users and how we will test a given PR against that plan.

We have a tricky situation because we have a mixture of (a) users reading content (that is, the quickstart instructions and the bootstrap script) directly from the website, with either `main` or a release (of which there is currently only one available, `v0.2.2`) selected in the menu at the top and (b) users downloading some release and using its content.  In particular, when viewing `main` on the website, a user sees the latest quickstart instructions, which say to use the latest bootstrap script, which in turn will use some release (which generally is older than the latest content).  A change to the quickstart or bootstrap script immediately starts affecting users, but they will still be using a release that has older content.  Conversely, tagging a new release jumps the release content that users will be using without changing the quickstart instructions or the bootstrap script.  Both kinds of changes thus have compatability issues.  It gets even more fun if we imagine users who have read the quickstart instructions and embedded them in the user's own scripting, and/or who have read the bootstrap script and copied and modified it into the user's own automation.

There is also the issue of evolving the syncer.  Currently the `kubectl kubestellar syncer-gen` plugin defaults to a fixed image reference for the syncer, `quay.io/kubestellar/syncer:v0.2.1`.  That will need to change in the future.  As with other references, source material in a released commit can not refer to itself unless we make a prior commit that refers to a release that does not (at that time) exist yet --- which seems bad.  OTOH, `make build` can build into executables a reference to the commit that they were built from (and in fact is already doing this for go programs).

An earlier and less comprehensive and admittedly incomplete take on this is in https://github.com/kcp-dev/edge-mc/blob/main/CONTRIBUTING.md#testing-locally .  Issue #422 is about the syncer part of this.

**Comment by MikeSpreitzer** [comment]
For evolving the default syncer image reference I can see three approaches.

In one approach the default for the syncer image reference has a tag holding the semantic version of a release, and every syncer change is followed by two releases.  The first following release creates a new syncer image, and the second updates the default in the syncer-gen plugin to refer to the new image.  This seems clunky.

In another approach the default for the syncer image reference has a tag holding the semantic version of a release, and every syncer change is followed by an anticipatory commit that updates the default in the syncer-gen plugin to use the expected semantic version of the upcoming release.  The image-defaulting behavior in this commit can not be tested until the anticipated release is made.  Not even in CI.  This seems bad.

In the third approach, the default for the syncer image reference has a tag holding a git commit hash --- the hash identifying the commit that the syncer-gen executable was built from.  This allows testing of the defaulting behavior in a commit before tagging a release.  But it _requires_ building the syncer image and pushing it to quay for every commit that is tested.  This can be a bit tedious, and creates image sprawl.  It would be safe to delete images that do not correspond to a release or a recent build (one that might still be being tested).

None of these seem great to me.  The third seems the least bad.

**Comment by MikeSpreitzer** [comment]
For consistency between the quickstart, bootstrap script, and most recent release contents one approach would be to give up on consistency between the latest website contents and the most recent release.  Make the quickstart instructions explicitly tell the reader that they are not promised to work unless the reader is looking at the quickstart in a regular release; here "regular" means to exclude releases that githib calls "pre-releases", which we reserve for testing.  Make the quickstart tell the reader to note the release that they are looking at, and use the bootstrap script from the same release.  Make these instructions explicitly tell the user that the quickstart instructions and bootstrap script can change in incompatible ways from one release to the next.  Since the user was also told to use a release-consistent set of things, that is not a problem; the important implication is that any automation that the user has built based on one release may have to change for a later one.

In this approach, between releases, we can make any changes we want to the quickstart and/or bootstrap script --- because no users will be using the revised versions yet.  To prepare for a release, we get the quickstart and bootstrap script to be what we hope is consistent with each other and the contents that will go into the upcoming release (ideally we would always maintain that, but temporary excursions are not a real problem).  Then we make that release, as a github "pre-release".  Now we can test everything together, except (a) the bits of script that fill in default values based on the latest regular release and possibly (b) the bits of functionality that provide the default value for the syncer image reference.  For the untestable behavior we just grit our teeth and hope for the best; we have reason to hope because this is a very small and isolated bit of functionality that is already working (for (a) anyway).  If testing passes then we tell github to promote the release from being a "pre-release" to a regular release (the latest).  If testing fails then we go back to making changes, to fix the problems; users have not been perturbed yet.

A downside of this approach is that it relies on users to do as they are told regarding picking a version and sticking to it.  Many projects to do this to a lesser degree.  You will often see "install" or "getting started" instructions that say to pick a release and download an archive for that release and the user's platform.  In our case it is one step beyond, the user has to consistently use a chosen release in picking the quickstart version to read and the bootstrap script version to use.  Or we could put a specific release semver in the part of the quickstart that holds the URL for the bootstrap script, so that we do not have to ask users to do an edit on their own.  That creates a self-reference problem: a commit can not hold source material with a literal release semver that refers to itself unless that commit is made before the release with that semver is tagged.  In this instance of the self-reference problem I think the anticipatory source solution is usable; we will not have CI that reads and follows the quickstart instructions (at least for a little while longer) and human testing is already supposed to be held off until a release is made.

Another downside of this approach is that no change to the quickstart or bootstrap can be delivered without a regular release.  See the PRs in flight right now that change the host port numbers used in the quickstart for an example of something we would want to deliver soon.

**Comment by MikeSpreitzer** [comment]
For consistency between the quickstart, bootstrap script, and most recent release contents another approach would be to give up on joint evolution and require almost complete everything-to-everything compatibility.  Every version of the quickstart, every version of the bootstrap script, and the most recent release have to work together.  I think that would quickly be seen as too strong a constraint.  We could back off a bit, and replace both of those "every" with "every recent" with some definition of "recent" and explicit admonition that using stuff older than that is not promised to work.  Remember that users looking at the website are _not_ atomically fetching the bootstrap script along with the quickstart instructions, so there has to be at least a little tolerance for version skew in there.

A downside of this is that it makes for a big test matrix.  Every PR to either the quickstart or the bootstrap script has to be _added_ to one dimension of the test matrix (or both if it changes both); dimensions get culled according to the "recent" criterion.  Similarly, making a new release requires testing that new release with all recent versions of the quickstart and all recent versions of the bootstrap script.

We could fix the test matrix to a modest (but still tedious --- we are talking about human testers!) size of `2*2*1` by defining "recent" to be the two latest versions.  That would effectively rule out users making any automation based on what they read unless that automation replaces both the quickstart and bootstrap; that jointness requirement may be tolerable.

**Comment by MikeSpreitzer** [comment]
My commends above were written under the misconception that the website contents are served from either the `main` branch or a release tag (depending on what is selected in the menu at the top).  The truth of the matter is more complicated.

**Comment by clubanderson** [comment]
I have added a config.ks_branch and config.ks_tag variable to the documentation which is being used in the quickstart guide and can be extended to use in the any other documentation (like the kubestellar-syncer and beyond).
